### PR TITLE
feat(obsy): add logic for configuring obsy

### PR DESCRIPTION
**Pull request by smuu**
## Overview

Testing undergoing.

See example repo: https://github.com/celestiaorg/knuu-example/pull/40

See lines here: https://github.com/celestiaorg/knuu-example/pull/40/files#diff-5a2277d213db3e35b592f1297547b1f2afe07ba9e46079754031e199b4433866R38-R52

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: pkg/k8s/k8s_ingress.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CreateIngress creates a new Ingress resource.
+func CreateIngress(namespace, name string, labels, annotations map[string]string, ingressClassName, host, path, pathType, serviceName string, servicePort int, tlsEnabled bool) error {
```
Comment (jrmanes) [inline pkg/k8s/k8s_ingress.go:11]: wdyt?
```suggestion
func CreateIngress(
	namespace,
	name string,
	labels, 
	annotations map[string]string,
	ingressClassName, 
	host,
	path,
	pathType,
	serviceName string,
	servicePort int,
	tlsEnabled bool
) error {
```

File: pkg/knuu/instance.go [hunk 2]
```diff
@@ -18,31 +18,44 @@ import (
...
+	cpuRequest               string
+	policyRules              []rbacv1.PolicyRule
+	livenessProbe            *v1.Probe
+	readinessProbe           *v1.Probe
+	startupProbe             *v1.Probe
+	files                    []*k8s.File
+	isSidecar                bool
+	parentInstance           *Instance
+	sidecars                 []*Instance
+	ingress                  *Ingress
```
Comment (jrmanes) [inline pkg/knuu/instance.go:46]: maybe is adding more complexity, did you consider splitting the obsy values in another struct?

File: pkg/knuu/instance_helper.go [hunk 3]
```diff
@@ -185,19 +185,16 @@ func (i *Instance) destroyPod() error {
...
-			if err != nil {
-				return fmt.Errorf("error patching service for instance '%s': %w", i.k8sName, err)
-			}
+	svc, _ := k8s.GetService(k8s.Namespace(), i.k8sName)
+	if svc == nil {
+		err := i.deployService()
+		if err != nil {
+			return fmt.Errorf("error deploying service for instance '%s': %w", i.k8sName, err)
+		}
+	} else if svc != nil {
```
Comment (jrmanes) [inline pkg/knuu/instance_helper.go:194]: just for curiosity
can `svc` be something different than `nil`? 
I wonder why the `else if svc != nil`

File: pkg/knuu/instance.go (hunk 2 above)
Comment (Bidon15) [reply pkg/knuu/instance.go:46]: Agreed with Jose here. Please consider avoiding `hero` structs or funcs. 
You can avoid either by duck-typing or you can get inspired by node repo with using Options / Builders pattern

File: pkg/knuu/instance_helper.go (hunk 3 above)
Comment (Bidon15) [reply pkg/knuu/instance_helper.go:194]: please consider looking to switch statements as I understood that there is a 3rd option for svc variable.

File: pkg/knuu/instance_otel.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+					Endpoint: fmt.Sprintf("localhost:%d", i.jaegerThriftCompactPort),
+				},
+				ThriftHTTP: JaegerThriftHTTP{
+					Endpoint: fmt.Sprintf("localhost:%d", i.jaegerThriftHttpPort),
+				},
+			},
+		}
+	}
+
+	return receivers
```
Comment (Bidon15) [inline pkg/knuu/instance_otel.go:253]: for readibility purposes, I highly recommend to split this into funcs for each receiver we are setting for the overall receiver struct.

File: pkg/knuu/instance_otel.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	if i.jaegerEndpoint != "" {
+		exporters.Jaeger = JaegerExporter{
+			Endpoint: i.jaegerEndpoint,
+			TLS: TLS{
+				Insecure: true,
+			},
+		}
+	}
```
Comment (Bidon15) [inline pkg/knuu/instance_otel.go:275]: same here for consistency.

**Review by Bidon15 — CHANGES_REQUESTED** [review]

File: pkg/knuu/instance.go (hunk 2 above)
Comment (smuu) [reply pkg/knuu/instance.go]: https://github.com/celestiaorg/knuu/pull/174/commits/0dd1a454e3b05f7cfdf2b6c5a62df15aa53194fd

File: pkg/knuu/instance_helper.go (hunk 3 above)
Comment (smuu) [reply pkg/knuu/instance_helper.go]: There is not 3rd option, good catch!
https://github.com/celestiaorg/knuu/pull/174/commits/4d826dc2b0b9f7f08eafc85a6ff7ebf561821dde

File: pkg/knuu/instance.go [hunk 6]
```diff
@@ -16,35 +16,7 @@ import (
...
-	livenessProbe            *v1.Probe
-	readinessProbe           *v1.Probe
-	startupProbe             *v1.Probe
-	files                    []*k8s.File
-	isSidecar                bool
-	parentInstance           *Instance
-	sidecars                 []*Instance
-	ingress                  *Ingress
-	externalDns              []string
+type ObsyConfig struct {
```
Comment (Bidon15) [inline pkg/knuu/instance.go:19]: [proposal][feature-request]: please make a dedicated issue that tackles refactoring of obsyconfig to more composable structs per type of stack we use for metrics/traces (otlp, prom, jaeger, pyro, etc.)

File: pkg/knuu/instance.go [hunk 7]
```diff
@@ -58,6 +30,38 @@ type Instance struct {
 	jaegerEndpoint           string
```
Comment (Bidon15) [inline pkg/knuu/instance.go:31]: please place jaegerEndpoint with other jaeger namings close to each other for readability. 
In addition, please make comment lines between each options

File: pkg/knuu/instance_otel.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+	CheckInterval string `yaml:"check_interval,omitempty"`
+}
+
+func (i *Instance) createOtelCollectorInstance() (*Instance, error) {
+	otelAgent, err := NewInstance("otel-agent")
+	if err != nil {
+		return nil, fmt.Errorf("error creating otel-agent instance: %w", err)
+	}
+
+	if err := otelAgent.SetImage("otel/opentelemetry-collector-contrib:0.71.0"); err != nil {
```
[the remark this answers is not in the document]
Comment (smuu) [reply pkg/knuu/instance_otel.go]: https://github.com/celestiaorg/knuu/pull/174/commits/6de40d7736b657b9a621f8ef8a4f440c3970a9a1

File: pkg/knuu/instance_helper.go [hunk 9]
```diff
@@ -334,44 +334,34 @@ func (i *Instance) cloneWithSuffix(suffix string) *Instance {
...
-		prometheusScrapeInterval: i.prometheusScrapeInterval,
-		jaegerGrpcPort:           i.jaegerGrpcPort,
-		jaegerThriftCompactPort:  i.jaegerThriftCompactPort,
-		jaegerThriftHttpPort:     i.jaegerThriftHttpPort,
-		otlpEndpoint:             i.otlpEndpoint,
-		otlpUsername:             i.otlpUsername,
-		otlpPassword:             i.otlpPassword,
-		jaegerEndpoint:           i.jaegerEndpoint,
+		name:                  i.name + suffix,
+		k8sName:               i.k8sName + suffix,
```
Comment (Bidon15) [inline pkg/knuu/instance_helper.go]: [proposal]: I would pass it for now, but we need a checker(issue to be created) that stops creating an instance if the suffix has wrong conventions. E.g. if the user passes a space symbol with a proper char set?

**Review by Bidon15 — CHANGES_REQUESTED** [review]
Close to being merged 👍

File: pkg/knuu/instance_otel.go (hunk 4 above)
Comment (smuu) [reply pkg/knuu/instance_otel.go]: https://github.com/celestiaorg/knuu/pull/174/commits/63b25751d3c9e9390afb8b154344d15d5396becb

File: pkg/knuu/instance_otel.go (hunk 5 above)
Comment (smuu) [reply pkg/knuu/instance_otel.go]: https://github.com/celestiaorg/knuu/pull/174/commits/63b25751d3c9e9390afb8b154344d15d5396becb

File: pkg/knuu/instance.go [hunk 10]
```diff
@@ -670,6 +692,67 @@ func (i *Instance) AddExternalDns(dns string) error {
 	return nil
 }
 
+// SetOtelEndpoint sets the OpenTelemetry endpoint for the instance
+// This function can only be called in the state 'Preparing' or 'Committed'
+func (i *Instance) SetOtelEndpoint(port int) error {
+	if !i.IsInState(Preparing, Committed) {
+		return fmt.Errorf("setting OpenTelemetry endpoint is only allowed in state 'Preparing' or 'Committed'. Current state is '%s'", i.state.String())
+	}
```
[the remark this answers is not in the document]
Comment (smuu) [reply pkg/knuu/instance.go]: https://github.com/celestiaorg/knuu/pull/174/commits/63959b939c2d2f960e0983eaf2e41b8270eef406

File: pkg/knuu/instance_helper.go (hunk 9 above)
Comment (smuu) [reply pkg/knuu/instance_helper.go]: https://github.com/celestiaorg/knuu/issues/15

File: pkg/knuu/instance.go (hunk 7 above)
Comment (smuu) [reply pkg/knuu/instance.go]: https://github.com/celestiaorg/knuu/pull/174/commits/943f4bebc89cf114a5059e693fbb825751faaddd

File: pkg/knuu/instance_helper.go [hunk 11]
```diff
@@ -185,19 +185,16 @@ func (i *Instance) destroyPod() error {
...
-			err := i.deployService()
-			if err != nil {
-				return fmt.Errorf("error deploying service for instance '%s': %w", i.k8sName, err)
-			}
-		} else if svc != nil {
-			err := i.patchService()
-			if err != nil {
-				return fmt.Errorf("error patching service for instance '%s': %w", i.k8sName, err)
-			}
+	svc, _ := k8s.GetService(k8s.Namespace(), i.k8sName)
```
[the remark this answers is not in the document]
Comment (smuu) [reply pkg/knuu/instance_helper.go]: Added proper handling: https://github.com/celestiaorg/knuu/pull/174/commits/c2a33a96ee2415476d9879a5936b8f8bde316fed

File: pkg/knuu/instance.go (hunk 6 above)
Comment (smuu) [reply pkg/knuu/instance.go:20]: https://github.com/celestiaorg/knuu/issues/178

File: pkg/knuu/instance_otel.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+	Exporters  []string `yaml:"exporters,omitempty"`
+	Processors []string `yaml:"processors,omitempty"`
+}
+
+type Processors struct {
+	Batch         Batch         `yaml:"batch,omitempty"`
+	MemoryLimiter MemoryLimiter `yaml:"memory_limiter,omitempty"`
+}
+
+type Batch struct{}
```
[the remark this answers is not in the document]
Comment (Bidon15) [reply pkg/knuu/instance_otel.go:136]: Understood. Thanks for pointing it out

File: pkg/knuu/instance_otel.go [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+			traces.Receivers = append(traces.Receivers, "jaeger")
+		}
+		if i.obsyConfig.otlpEndpoint != "" {
+			traces.Exporters = append(traces.Exporters, "otlphttp")
+		}
+		if i.obsyConfig.jaegerEndpoint != "" {
+			traces.Exporters = append(traces.Exporters, "jaeger")
+		}
+		pipelines.Traces = traces
+	}
```
Comment (Bidon15) [inline pkg/knuu/instance_otel.go:335]: Let's refactor this into prepare metrics and traces private funcs. 
You can make also make an extra-mile either now or in next PRs where we can squash those if statements + appends

**Review by Bidon15 — APPROVED** [review]

**Pull request merged**